File: packages/sdk-python/doubleoh/tools.py
"""The loop as agent tools — plain functions, which is what most frameworks now want.

The OpenAI Agents SDK, AutoGen, and Pydantic AI all accept ordinary Python callables as
tools, reading the name, signature, and docstring. So the base adapter is exactly that:
``build_tools(client)`` returns three well-documented functions, and those three
frameworks need nothing else::

    from doubleoh import DoubleOh
    from doubleoh.tools import build_tools

    tools = build_tools(DoubleOh(api_key=...))

    # OpenAI Agents SDK:   Agent(tools=[function_tool(f) for f in tools])
    # AutoGen:             AssistantAgent(tools=list(tools))
    # Pydantic AI:         Agent(tools=list(tools))

The functions return STRINGS written for a model to read: a tool result is context, not
an API payload, and "A colleague has done this before: …" steers a model where a JSON
blob does not. LangChain/LangGraph, CrewAI, and LlamaIndex have their own modules that
wrap these same functions in each framework's tool class.
"""

from __future__ import annotations

from typing import Callable

from . import DoubleOh, DoubleOhError, LearnedSkill


def _skill_lines(skill: LearnedSkill) -> str:
    return (
        f'Skill "{skill.name}" (worked {skill.times_worked}×, failed {skill.times_failed}×):\n'
        f"{skill.description}\n{skill.instructions}"
    )


def build_tools(client: DoubleOh) -> tuple[Callable, Callable, Callable]:
    """The three tools, bound to one client, in the order an agent uses them."""

    def doubleoh_skills_for(task: str) -> str:
        """BEFORE attempting a task you have failed at before, ask what has been learned.

        Returns step-by-step procedures compiled from past human fixes, with how often
        each has worked. Follow a returned procedure instead of improvising.

        Args:
            task: The task, in one sentence. Phrase it the same way each time — the
                phrasing is part of the skill's identity.
        """
        skills = client.skills_for(task)
        if not skills:
            return (
                "Nothing has been learned about this task yet. Proceed, and if you get "
                "stuck, use doubleoh_request_fix."
            )
        rendered = "\n\n".join(_skill_lines(skill) for skill in skills)
        return f"A colleague has done this before. Follow these steps:\n\n{rendered}"

    def doubleoh_request_fix(url: str, task: str) -> str:
        """When you are STUCK on a web page, ask a human to fix it.

        A login wall, a changed layout, a step you cannot complete — a person fixes it
        once in a live browser and the fix becomes a skill the whole fleet reuses. If the
        fleet already knows this wall, the learned procedure is returned instantly
        instead and nobody is paged: follow it and retry. Do not call this for tasks you
        have not attempted.

        Args:
            url: The public http(s) page you are stuck on.
            task: What you were trying to do, in one sentence (max 500 characters).
        """
        try:
            fix = client.request_fix(url=url, task=task)
        except DoubleOhError as error:
            # A refusal is an answer the model should read, not an exception to crash on.
            return f"DoubleOh refused ({error.status}): {error}"
        if fix.deflected and fix.skill:
            return (
                "No human was needed — this wall is already known. Follow this procedure "
                f"and retry:\n\n{_skill_lines(fix.skill)}\n\n"
                "If it still fails, call doubleoh_request_fix again: the second ask "
                "reaches a human."
            )
        if fix.duplicate_of:
            blocked = fix.blocked_runs or "several"
            return (
                f"A human has already been asked about this exact wall ({blocked} runs "
                "are blocked on it). Do not ask again; move to other work or wait."
            )
        return (
            f"A human has been asked to fix this. Fix link (already sent to the team if "
            f"a webhook is configured): {fix.fix_url}. Move on to other work; the fix "
            "becomes a reusable skill once a person completes it."
        )

    def doubleoh_report_skill(name: str, worked: bool) -> str:
        """AFTER following a skill, report whether it worked.

        This keeps the skill's track record honest; a skill that keeps failing stops
        being served. Always call this when you used a skill.

        Args:
            name: The skill's name, exactly as returned.
            worked: True if following the skill completed the task.
        """
        client.report_skill(name, worked)
        return "Recorded. Thank you — the skill's track record is what the next agent trusts."

    return doubleoh_skills_for, doubleoh_request_fix, doubleoh_report_skill

File: packages/sdk/src/tools.ts
/**
 * The loop as agent tools, framework-agnostic.
 *
 * Every current tool-calling convention (MCP, Vercel AI SDK, Mastra, OpenAI function calling) wants
 * the same three things: a name, a JSON schema for the arguments, and something to execute. So that
 * is what this exports, and each framework wraps it in one line instead of this package depending
 * on all of them.
 *
 * The executors return STRINGS, written for a model to read: a tool result is context, not an API
 * payload. Each result ends by naming the next tool to call, because the model that reads it is
 * deciding what to do next, and the descriptions teach the whole loop, not one call.
 */

import type { DoubleOh, LearnedSkill } from "./index.js";

export type DoubleOhTool = {
  name: string;
  description: string;
  /** JSON Schema for the arguments, the lingua franca every framework accepts. */
  parameters: {
    type: "object";
    properties: Record<string, { type: string; description: string }>;
    required: string[];
  };
  execute: (args: Record<string, unknown>) => Promise<string>;
};

const LOOP =
  "The loop: doubleoh_skills_for before a task you have failed at before; doubleoh_request_fix once " +
  "when stuck; doubleoh_wait_for_fix while a person works; doubleoh_report_skill after following any skill.";

/** The skill as context: the server's hint when it has one (boundary first), else the same shape built here. */
function skillLines(skill: LearnedSkill): string {
  if (skill.systemHint) return skill.systemHint;
  const authority = skill.authority;
  const boundary = authority
    ? `Requires a human: ${authority.requiresHuman.join("; ") || "nothing stated"}. Never: ${authority.never.join("; ") || "nothing stated"}.`
    : "Boundary not stated: stop and ask a person before any credential, payment or approval.";
  return (
    `Skill "${skill.name}" (worked ${skill.timesWorked}, failed ${skill.timesFailed}). ${skill.description}\n` +
    `Authority. ${boundary}\nSteps:\n${skill.instructions}`
  );
}

/**
 * The four tools, bound to one client.
 *
 * Vercel AI SDK:  aisdk `tool({ description, inputSchema: jsonSchema(t.parameters), execute: t.execute })`
 * OpenAI:         `{ type: "function", function: { name: t.name, description: …, parameters: … } }`
 * MCP:            served directly by doubleoh-mcp.
 */
export function doubleohTools(client: DoubleOh): DoubleOhTool[] {
  return [
    {
      name: "doubleoh_skills_for",
      description:
        "BEFORE attempting a task you have failed at before, ask what the fleet has learned about it " +
        "on this page. Returns procedures compiled from past human fixes, each with its record and " +
        "its authority boundary: what still needs a person, what you must never do. Follow a returned " +
        `procedure instead of improvising, and stay inside its boundary. ${LOOP}`,
      parameters: {
        type: "object",
        properties: {
          task: {
            type: "string",
            description:
              "The task, in one sentence. Phrase it the same way each time; the phrasing is part of the skill's identity.",
          },
          url: {
            type: "string",
            description:
              "The page you are on. Scopes the answer to the site the skill was learned on. Pass it whenever you have it.",
          },
        },
        required: ["task"],
      },
      execute: async (args) => {
        const url = typeof args.url === "string" ? args.url : undefined;
        const skills = await client.skillsFor(String(args.task ?? ""), { url });
        if (skills.length === 0) {
          return (
            "Nothing has been learned about this task on this site yet. Attempt it. " +
            "If you hit a login, a code, a dialog or an approval you cannot pass, call doubleoh_request_fix once."
          );
        }
        return `A colleague has done this before. Follow the first procedure, inside its boundary:\n\n${skills
          .map(skillLines)
          .join(
            "\n\n",
          )}\n\nWhen done, call doubleoh_report_skill with the skill's name and whether it worked.`;
      },
    },
    {
      name: "doubleoh_request_fix",
      description:
        "When you are STUCK on a page (a login wall, a one-time code, a consent dialog, an approval, a " +
        "changed layout) ask for a person. Call it ONCE per wall; retrying the action first, or asking " +
        "repeatedly, locks accounts and pages people for nothing. If the fleet already knows this wall, " +
        "the learned procedure is returned instantly instead and nobody is paged: follow it and retry. " +
        `Otherwise a person takes over the live session; call doubleoh_wait_for_fix to know when. ${LOOP}`,
      parameters: {
        type: "object",
        properties: {
          url: {
            type: "string",
            description:
              "The page you are stuck on. Public http(s), or a host your team allowlisted for its own runtime.",
          },
          task: {
            type: "string",
            description:
              "What you were trying to do, in one sentence (max 500 characters), phrased as in doubleoh_skills_for.",
          },
        },
        required: ["url", "task"],
      },
      execute: async (args) => {
        const fix = await client.requestFix({
          url: String(args.url ?? ""),
          task: String(args.task ?? ""),
        });
        if (fix.deflected && fix.skill) {
          return (
            "No person was needed: this wall is already known. Follow this procedure inside its boundary and retry:\n\n" +
            skillLines(fix.skill) +
            "\n\nThen call doubleoh_report_skill. If it still fails, call doubleoh_request_fix again; the second ask reaches a person."
          );
        }
        if (fix.duplicateOf) {
          return (
            `A person has already been asked about this exact wall (${fix.blockedRuns ?? "several"} runs are blocked on it; wall ${fix.wallId ?? fix.duplicateOf}). ` +
            `Do not ask again. Call doubleoh_wait_for_fix with id ${fix.duplicateOf}, or move to other work.`
          );
        }
        return (
          `A person has been asked (intervention ${fix.id}, wall ${fix.wallId ?? "unknown"}). ` +
          `Fix link, already sent to the team if a webhook is configured: ${fix.fixUrl}. ` +
          `Typical time: about ${Math.round((fix.expectedSeconds ?? 300) / 60)} minutes. ` +
          `Call doubleoh_wait_for_fix with id ${fix.id} to be told when it is done, and do other work meanwhile.`
        );
      },
    },
    {
      name: "doubleoh_wait_for_fix",
      description:
        "After doubleoh_request_fix returned an intervention id, wait for the person to finish. One " +
        "call holds up to 55 seconds and returns as soon as the status changes; call it again while " +
        "the answer says still open. When resolved, call doubleoh_skills_for with the same task and " +
        `url to get the procedure the fix produced. ${LOOP}`,
      parameters: {
        type: "object",
        properties: {
          id: {
            type: "string",
            description: "The intervention id from doubleoh_request_fix.",
          },
        },
        required: ["id"],
      },
      execute: async (args) => {
        const status = await client.intervention(String(args.id ?? ""), {
          waitSeconds: 55,
        });
        if (status.status === "open") {
          return `Still open; a person has not finished yet. Call doubleoh_wait_for_fix again with id ${status.id}, or continue other work.`;
        }
        if (status.status === "resolved") {
          return (
            `Resolved by a person.${status.skillName ? ` A skill was compiled: "${status.skillName}".` : " No skill was compiled from it."} ` +
            "Retry the task now; call doubleoh_skills_for with the same task and url first to get the procedure, and stay inside its boundary."
          );
        }
        return `The fix ended as "${status.status}" without a result. Call doubleoh_request_fix again with the same url and task to reach a person.`;
      },
    },
    {
      name: "doubleoh_report_skill",
      description:
        "AFTER following a skill from doubleoh_skills_for or a deflection, report whether it worked. " +
        "This keeps the skill's track record honest; a skill that keeps failing stops being served. " +
        `Always call this when you used a skill, whether it worked or not. ${LOOP}`,
      parameters: {
        type: "object",
        properties: {
          name: {
            type: "string",
            description: "The skill's name, exactly as returned.",
          },
          worked: {
            type: "boolean",
            description: "true if following the skill completed the task.",
          },
        },
        required: ["name", "worked"],
      },
      execute: async (args) => {
        await client.reportSkill(String(args.name ?? ""), Boolean(args.worked));
        return "Recorded. The skill's track record is what the next agent trusts. Continue with your task.";
      },
    },
  ];
}
